runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java

```diff
@@ -298,6 +303,11 @@ private void shutdownIfNecessary(State newState) {
     } catch (final RuntimeException re) {
       errors.add(re);
     }
+    try {
+        metricsExecutor.shutdown();
```

Comment:
It's worth filing a follow-up jira to try to minimize the surface area of things the `ExecutorServiceParallelExecutor` needs to know about to shut down.

Reply:
https://issues.apache.org/jira/browse/BEAM-4039